src/shogun/lib/any.h

```diff
@@ -138,37 +137,40 @@ namespace shogun
 		};
 
 		template <class T>
-		auto compare_impl(by_default, T& lhs, T& rhs) = delete;
+		auto compare_impl(by_default, const T& lhs, const T& rhs) = delete;
 
 		template <class T>
-		auto compare_impl(general, T& lhs, T& rhs) -> decltype(lhs == rhs)
+		bool compare_impl_eq(const T& lhs, const T& rhs)
 		{
-			SG_SDEBUG("Comparing using lhs==rhs.\n");
 			return lhs == rhs;
 		}
+		template <>
+		bool compare_impl_eq(const float32_t& lhs, const float32_t& rhs);
+		template <>
+		bool compare_impl_eq(const float64_t& lhs, const float64_t& rhs);
+		template <>
+		bool compare_impl_eq(const floatmax_t& lhs, const floatmax_t& rhs);
+		template <>
+		bool compare_impl_eq(const complex128_t& lhs, const complex128_t& rhs);
 
-		auto compare_impl(more_important, float32_t& lhs, float32_t& rhs)
-		    -> bool;
-		auto compare_impl(more_important, float64_t& lhs, float64_t& rhs)
-		    -> bool;
-		auto compare_impl(more_important, floatmax_t& lhs, floatmax_t& rhs)
-		    -> bool;
-		auto compare_impl(more_important, complex128_t& lhs, complex128_t& rhs)
-		    -> bool;
+		template <class T>
+		auto compare_impl(general, const T& lhs, const T& rhs)
+		    -> decltype(lhs == rhs)
+		{
+			return compare_impl_eq(lhs, rhs);
+		}
 
 		template <class T>
-		auto compare_impl(more_important, T& lhs, T& rhs)
+		auto compare_impl(more_important, const T& lhs, const T& rhs)
 		    -> decltype(lhs.equals(rhs))
 		{
-			SG_SDEBUG("Comparing using lhs.equals(rhs).\n");
```

Comment:
I think these are incredibly useful.
Can't you move the implementation to the .cpp and keep them?

Reply:
Yes we can. Let me do that in other PR.